pkg/billing/azureblobstoragebilling.go

```diff
@@ -0,0 +1,158 @@
+package billing
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob"
+	"github.com/lithammer/shortuuid"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+type AzureAuthType string
+
+const (
+	// AzureAuthTypeTests is used in tests
+	// It uses well-known storage account and key.
+	// See https://learn.microsoft.com/en-us/azure/storage/common/storage-use-azurite
+	AzureAuthTypeTests AzureAuthType = "tests"
+	// AzureAuthTypeDefault is for pods running in Azure Kubernetes.
+	// Make sure you have provisioned Role and you have Managed Identity.
+	AzureAuthTypeDefault AzureAuthType = "default"
+)
+
+type AzureAuthSharedKey struct {
+	AccountName string `json:"accountName"`
+	AccountKey  string `json:"accountKey"`
+}
+
+type AzureBlobStorageClientConfig struct {
+	AuthType  AzureAuthType       `json:"authType"`
+	SharedKey *AzureAuthSharedKey `json:"sharedKey"`
+	// In Azure a Container is close to a bucket in AWS S3
+	Container string `json:"container"`
+	// Files will be created with name starting with PrefixInContainer
+	PrefixInContainer string `json:"prefixInContainer"`
+	// Example Endpoint: "https://MYSTORAGEACCOUNT.blob.core.windows.net/"
+	Endpoint string `json:"endpoint"`
+
+	//
+	// Unexported attributes follow this comment.
+	//
+
+	// Use generateKey for tests.
+	// Otherwise, keep empty.
+	generateKey func() string
+	// trackProgress is useful in tests, it's invoked when SDK sends blobs to Azure.
+	// Otherwise keep empty.
+	trackProgress func(bytesTransferred int64)
+}
+
+type AzureError struct {
+	Err error
+}
+
+func (e AzureError) Error() string {
+	return fmt.Sprintf("S3 error: %s", e.Err.Error())
+}
+
+func (e AzureError) Unwrap() error {
+	return e.Err
+}
+
+type AzureClient struct {
+	cfg AzureBlobStorageClientConfig
+	c   *azblob.Client
+}
+
+func (c AzureClient) LogFields() zap.Field {
+	return zap.Inline(zapcore.ObjectMarshalerFunc(func(enc zapcore.ObjectEncoder) error {
+		enc.AddString("container", c.cfg.Container)
+		enc.AddString("prefixInContainer", c.cfg.PrefixInContainer)
+		enc.AddString("endpoint", c.cfg.Endpoint)
+		return nil
+	}))
+}
+
+func (c AzureClient) generateKey() string {
+	if c.cfg.generateKey != nil {
+		return c.cfg.generateKey()
```

Comment:
Instead of this, I'd add an argument `now time.Time`, and verified in tests that the key is what it is supposed to look like for the hardcoded timestamp.

Reply:
I wanted to test that I can read what I wrote, not the format of the `AzureClient.generateKey()`. If you wish to test the format we can have another test on that.

Reply:
I am just concerned about adding the field which is only used in tests.

Perhaps, you could extract the actual key generation into a standalone `func generateKey`, which would be the default value for the `cfg.generateKey`, also achieving the de-duplication with S3 code.

Reply:
> I am just concerned about adding the field which is only used in tests.

It's not exposed to users of the package.

> Perhaps, you could extract the actual key generation into a standalone func generateKey, which would be the default value for the cfg.generateKey, also achieving the de-duplication with S3 code.

Yeah, thought about this too, and did that.

Reply:
> Yeah, thought about this too, and did that.

I imagined one more step - setting `keyTemplate` as the default value for `client.cfg.generateKey` (it will have to accept an argument), and overwriting it in tests, and always calling `cfg.generateKey`. This way there will be no test-only code in the main file.

Reply:
sure.